diff --git a/gdb/infrun.c b/gdb/infrun.c
--- a/gdb/infrun.c
+++ b/gdb/infrun.c
@@ -2193,7 +2193,7 @@ static const char *const scheduler_enums[] = {
   schedlock_replay,
   NULL
 };
-static const char *scheduler_mode = schedlock_replay;
+static const char *scheduler_mode = schedlock_step;
 static void
 show_scheduler_mode (struct ui_file *file, int from_tty,
 		     struct cmd_list_element *c, const char *value)
