diff --git a/gdb/breakpoint.c b/gdb/breakpoint.c
--- a/gdb/breakpoint.c
+++ b/gdb/breakpoint.c
@@ -11904,6 +11904,8 @@ update_global_location_list (enum ugll_insert_mode insert_mode)
 		 traps we can no longer explain.  */
 
 	      old_loc->events_till_retirement = 3 * (thread_count () + 1);
+	      /* Red Hat Bug 590623.  */
+	      old_loc->events_till_retirement *= 10;
 	      old_loc->owner = NULL;
 
 	      moribund_locations.push_back (old_loc);
